src/Orleans.Runtime/Placement/Rebalancing/ActiveRebalancerGrain.cs

```diff
@@ -0,0 +1,723 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using System.Runtime.CompilerServices;
+using System.Diagnostics;
+using System.Collections.Immutable;
+using Orleans.Concurrency;
+using Orleans.Serialization.Invocation;
+using Orleans.Core.Internal;
+using Orleans.Runtime.Configuration.Options;
+using System.Data;
+using System.Diagnostics.CodeAnalysis;
+using Orleans.Placement.Rebalancing;
+using System.Threading;
+using Orleans.Placement;
+using Orleans.Internal;
+using System.Collections;
+
+namespace Orleans.Runtime.Placement.Rebalancing;
+
+// See: https://www.microsoft.com/en-us/research/wp-content/uploads/2016/06/eurosys16loca_camera_ready-1.pdf
+
+#nullable enable
+
+[KeepAlive]
+[PreferLocalPlacement]
+[MayInterleave(nameof(MayInterleave))]
+[GrainType(IActiveRebalancerGrain.TypeName)]
+internal sealed class ActiveRebalancerGrain : Grain, IActiveRebalancerGrain
+{
+    private readonly ILogger _logger;
+    private readonly IInternalGrainFactory _grainFactory;
+    private readonly IImbalanceToleranceRule _toleranceRule;
+    private readonly ActivationDirectory _activationDirectory;
+    private readonly ActiveRebalancingOptions _options;
+   
+    private static SiloAddress? _currentExchangeSilo;
+
+    private TaskCompletionSource? _unblocker;
+    private FrequencySink _frequencySink;
+    private DateTimeOffset? _lastExchangeTime;
+
+    [AllowNull] private IDisposable? _timer;
```

Comment:
Nit: 
```suggestion
    private IDisposable? _timer;
```
`null` is already allowed via `?`.